src/core/adapters/personalization.ts

```diff
@@ -0,0 +1,126 @@
+        break;
+      case Actions.SELECT_REFINEMENT:
+        const refinement = Selectors.refinementCrumb(state, payload.navigationId, payload.index);
+        field = refinement.field;
+        value = refinement.value;
+        break;
+      default:
+    }
+
+    return { field, value };
+  };
+
+  export const generateNewBias = () => ({
+    lastUsed: Math.floor(Date.now() / 1000)
+  });
+
+  // tslint:disable-next-line max-line-length
+  export const transformToBrowser = (state: Store.Personalization.Biasing, reducerKey: string): BrowserStorageState => ({
+    allIds: state.allIds ? state.allIds.map(({ variant, key }) => ({
+      variant,
+      key,
+      ...state.byId[variant][key]
+    })) : []
+  });
+
+  // tslint:disable-next-line max-line-length
+  export const transformFromBrowser = (incomingState: BrowserStorageState, state: Store.State): Store.Personalization.Biasing => {
+    const config = Selectors.config(state);
+    const olderThanTime = Math.floor(Date.now() / 1000) - DAYS_IN_SECONDS * ConfigAdapter.extractRealTimeBiasingExpiry(config);
+    const filteredIncomingState = incomingState.allIds.filter((element) => element.lastUsed >= olderThanTime);
+    let allIds = [];
+    const byId = {};
+    filteredIncomingState.forEach(({ variant, key, lastUsed }) => {
+      allIds.push({ variant, key });
+      byId[variant] = { ...byId[variant], [key]: { lastUsed } };
+    });
+    Object.keys(byId).forEach((key) => {
+      allIds = pruneBiases(allIds, key, Object.keys(byId[key]).length, config.personalization.realTimeBiasing);
+    }); // filter allIds for each variant
+    return {
+      allIds,
+      byId
+    };
+  };
+
+  export const convertBiasToSearch = (state: Store.State) => {
+    const allIds = Selectors.realTimeBiasesAllIds(state);
+    const config = Selectors.config(state).personalization.realTimeBiasing;
+    const selectedRefinements = Selectors.selectedRefinements(state);
+
+    return allIds.filter(({ variant, key }) => {
+      return !selectedRefinements.find(({ navigationName, type, value }) => {
+        return type === 'Value' && navigationName === variant && key && value === key;
+      });
+    }).map(({ variant, key }) => ({
+      name: variant,
+      content: key,
+      strength: (config.attributes[variant] && config.attributes[variant].strength) || config.strength
+    }));
+  };
+
+  // tslint:disable-next-line max-line-length
+  export const pruneBiases = (allIds: Store.Personalization.BiasKey[], variant: string, variantCount: number, config: Configuration.Personalization.RealTimeBiasing) => {
+    if (config.attributes[variant] && (variantCount >= (config.attributes[variant].maxBiases))) {
+      for (let i = allIds.length - 1; i >= 0; i--) {
+        if (allIds[i].variant === variant) {
+          return [...allIds.slice(0, i), ...allIds.slice(i + 1)];
+        }
+      }
+    }
+    if (allIds.length > config.maxBiases) {
+      return allIds.slice(0, config.maxBiases);
```

Comment:
is this in order to shallow clone the array? could be `[...config.maxBiases]` if you really need that

Reply:
We need to shallow clone the first n elements of allIds. Is there a way to do that with a spread?